File: backend-go/calendarservice/calendarservice.go
package calendarservice

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"time"

	"github.com/coreyk/piinky/backend-go/models"
	"golang.org/x/oauth2/google"
	calendar "google.golang.org/api/calendar/v3"
	"google.golang.org/api/option"
)

type Service struct {
	calendarSvc    *calendar.Service
	calendarConfig models.CalendarConfig
	now            func() time.Time // For testing purposes
}

func NewService() (*Service, error) {
	// Get the calendar config filename from the environment variable
	calendarConfigFile := os.Getenv("GOOGLE_CALENDAR_CONFIG_FILE")
	if calendarConfigFile == "" {
		calendarConfigFile = "../google_calendar_config.json"
		fmt.Println("GOOGLE_CALENDAR_CONFIG_FILE is not set in the .env file, using default: ", calendarConfigFile)
	}

	calendarConfigBytes, err := os.ReadFile(calendarConfigFile)
	if err != nil {
		return nil, fmt.Errorf("failed to read calendar config: %v", err)
	}

	var calendarConfig models.CalendarConfig
	if err := json.Unmarshal(calendarConfigBytes, &calendarConfig); err != nil {
		return nil, fmt.Errorf("failed to parse calendar config: %v", err)
	}

	calendarCredentialsFile := os.Getenv("GOOGLE_CALENDAR_CREDENTIALS_FILE")
	if calendarCredentialsFile == "" {
		calendarCredentialsFile = "../google_credentials.json"
		fmt.Println("GOOGLE_CALENDAR_CREDENTIALS_FILE is not set in the .env file, using default: ", calendarCredentialsFile)
	}

	calendarCredentialsBytes, err := os.ReadFile(calendarCredentialsFile)
	if err != nil {
		return nil, fmt.Errorf("failed to read credentials: %v", err)
	}

	calendarCreds, err := google.JWTConfigFromJSON(calendarCredentialsBytes, calendar.CalendarReadonlyScope)
	if err != nil {
		return nil, fmt.Errorf("failed to parse credentials: %v", err)
	}

	calendarSvc, err := calendar.NewService(context.Background(), option.WithHTTPClient(calendarCreds.Client(context.Background())))
	if err != nil {
		return nil, fmt.Errorf("failed to create calendar service: %v", err)
	}

	return &Service{
		calendarSvc:    calendarSvc,
		calendarConfig: calendarConfig,
		now:            time.Now,
	}, nil
}

// SetNowFunc allows setting a custom time function for testing
func (s *Service) SetNowFunc(f func() time.Time) {
	s.now = f
}

func (s *Service) HandleGetCalendar(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	// Get current time in local timezone
	today := s.now().In(time.Local)

	// Calculate the start of the week based on the configuration
	var weekStart time.Time
	if s.calendarConfig.StartOnSunday {
		// If today is Sunday (0), we want to stay on this Sunday
		daysToSubtract := int(today.Weekday())
		if daysToSubtract == 0 {
			weekStart = today
		} else {
			weekStart = today.AddDate(0, 0, -daysToSubtract)
		}
	} else {
		// For Monday start: if today is Monday (1), we stay on Monday
		daysToSubtract := int(today.Weekday())
		if daysToSubtract == 0 { // Sunday
			daysToSubtract = 6
		} else {
			daysToSubtract--
		}
		weekStart = today.AddDate(0, 0, -daysToSubtract)
	}

	// Ensure we start at beginning of day in local time
	weekStart = time.Date(weekStart.Year(), weekStart.Month(), weekStart.Day(), 0, 0, 0, 0, time.Local)

	timeMin := weekStart.Format(time.RFC3339)
	timeMax := today.AddDate(0, 0, s.calendarConfig.NumberOfWeeks*7).Format(time.RFC3339)

	var events []*calendar.Event
	for _, cal := range s.calendarConfig.Calendars {
		result, err := s.calendarSvc.Events.List(cal.CalendarID).
			TimeMin(timeMin).
			TimeMax(timeMax).
			SingleEvents(true).
			OrderBy("startTime").
			Do()
		if err != nil {
			http.Error(w, fmt.Sprintf("Failed to fetch calendar events: %v", err), http.StatusInternalServerError)
			return
		}

		for _, event := range result.Items {
			event.ColorId = cal.Color
			events = append(events, event)
		}
	}

	response := map[string]interface{}{
		"startDate":     timeMin,
		"endDate":       timeMax,
		"numberOfWeeks": s.calendarConfig.NumberOfWeeks,
		"startOnSunday": s.calendarConfig.StartOnSunday,
		"events":        events,
	}

	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(response); err != nil {
		http.Error(w, fmt.Sprintf("Failed to encode response: %v", err), http.StatusInternalServerError)
		return
	}
}

File: backend/piinky_backend/calendar_service.py
"""
This module provides the CalendarService class for interacting with Google Calendar API,
loading configuration, and retrieving calendar events.
"""

from datetime import datetime, timedelta
import json
from fastapi import HTTPException
from google.oauth2.service_account import Credentials
from googleapiclient.discovery import build

class CalendarService:
    """
    This class handles interactions with the Google Calendar API,
    including loading configuration and retrieving calendar events.
    """

    def __init__(self, credentials_file, config_file):
        self.load_config(credentials_file, config_file)

    def load_config(self, credentials_file, config_file):
        """
        Load configuration from the specified config file and set up
        Google Calendar credentials from the provided credentials file.
        """
        with open(config_file, 'r', encoding='utf-8') as f:
            config = json.load(f)
        self.calendars = config["calendars"]
        self.start_on_sunday = config["start_on_sunday"]
        self.number_of_weeks = config["number_of_weeks"]
        self.credentials = Credentials.from_service_account_file(credentials_file)
        self.service = build("calendar", "v3", credentials=self.credentials)

    async def get_calendar_data(self):
        """
        Retrieve calendar data for the current week, including events
        from configured calendars.
        """
        try:
            today = datetime.now().astimezone()
            if self.start_on_sunday:
                week_start = today - timedelta(days=((today.weekday() + 1) % 7))
            else:
                week_start = today - timedelta(days=today.weekday())

            end_date = week_start + timedelta(weeks=self.number_of_weeks)

            events = await self.get_events(week_start.isoformat(), end_date.isoformat())
            return {
                "startDate": week_start.isoformat(),
                "endDate": end_date.isoformat(),
                "numberOfWeeks": self.number_of_weeks,
                "startOnSunday": self.start_on_sunday,
                "events": events
            }
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e)) from e

    async def get_events(self, start_date: str, end_date: str) -> list:
        """
        Retrieve events from configured calendars within the specified date range.

        Args:
            start_date (str): The start date in ISO format.
            end_date (str): The end date in ISO format.

        Returns:
            list: A list of events from the calendars.
        """
        try:
            all_events = []
            for calendar in self.calendars:
                calendar_id = calendar['calendar_id']
                events_result = self.service.events().list(
                    calendarId=calendar_id,
                    timeMin=start_date,
                    timeMax=end_date,
                    singleEvents=True,
                    orderBy='startTime'
                ).execute()

                # Add calendar color to each event
                events = events_result.get('items', [])
                for event in events:
                    event['colorId'] = calendar.get('color', '#000')
                all_events.extend(events)

            return all_events
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e)) from e
